src/database/import.ts

```diff
@@ -0,0 +1,166 @@
+import * as clc from "colorette";
+import * as stream from "stream";
+import pLimit from "p-limit";
+
+import { URL } from "url";
+import { Client } from "../apiv2";
+import { FirebaseError } from "../error";
+
+const JSONStream = require("JSONStream");
+
+const MAX_CHUNK_SIZE = 1024 * 1024;
+const CONCURRENCY_LIMIT = 5;
+
+type Data = {
+  json: { [key: string]: any } | string | number | boolean;
+  pathname: string;
+};
+
+type ChunkedData = {
+  chunks: Data[] | null;
+  size: number;
+};
+
+/**
+ * Imports JSON data to a given RTDB instance.
+ *
+ * The data is parsed and chunked into subtrees of ~1 MB, to be subsequently written in parallel.
+ */
+export default class DatabaseImporter {
+  private jsonPath: string;
+  private client: Client;
+  private limit = pLimit(CONCURRENCY_LIMIT);
+
+  constructor(
+    private dbUrl: URL,
+    private inStream: NodeJS.ReadableStream,
+    dataPath: string,
+    private chunkSize = MAX_CHUNK_SIZE
+  ) {
+    this.jsonPath = this.computeJsonPath(dataPath);
+    this.client = new Client({ urlPrefix: dbUrl.origin, auth: true });
+  }
+
+  /**
+   * Writes the chunked data to RTDB. Any existing data at the specified location will be overwritten.
+   */
+  async execute(): Promise<any> {
+    await this.checkLocationIsEmpty();
+    return this.readAndWriteChunks(this.inStream);
+  }
+
+  private async checkLocationIsEmpty(): Promise<void> {
+    const response = await this.client.request({
+      method: "GET",
+      path: this.dbUrl.pathname + ".json",
+      queryParams: { shallow: "true" },
+    });
+
+    if (response.body) {
+      throw new FirebaseError(
+        "Importing is only allowed for an empty location. Delete all data by running " +
+          clc.bold(`firebase database:remove ${this.dbUrl.pathname} --disable-triggers`) +
+          ", then rerun this command.",
+        { exit: 2 }
+      );
+    }
+  }
+
+  private readAndWriteChunks(inStream: NodeJS.ReadableStream): Promise<any> {
+    const { dbUrl } = this;
+    const chunkData = this.chunkData.bind(this);
+    const writeChunk = this.writeChunk.bind(this);
+    const getJoinedPath = this.joinPath.bind(this);
+
+    const readChunks = new stream.Transform({ objectMode: true });
+    readChunks._transform = function (chunk: { key: string; value: any }, _, done) {
```

Comment:
Very cool. I am reading its docs: https://nodejs.org/api/stream.html#implementing-a-readable-stream

Thinking about how to control the size of the `chunk` emitted.
Shall we set `highWaterMark` higher to have larger chunk size?

Reply:
Not sure. We're in `objectMode`, so `highWaterMark` defaults to 16 (objects, not bytes). It's hard to know how many objects we should buffer.

Reply:
Can you experiment with a large file and log each request's size?

If it ends up sending one request for lots of tiny locations, it would be very slow.
We want to make sure the batch size is reasonable large for this command to be efficient and useful.